allennlp/data/dataset_readers/dataset_reader.py

```diff
@@ -233,25 +267,28 @@ def _read(self, file_path: str) -> Iterable[Instance]:
         raise NotImplementedError
 
     def _instances_from_cache_file(self, cache_filename: str) -> Iterable[Instance]:
-        # Try to acquire a lock just to make sure another process isn't in the middle
-        # of writing to the cache.
-        cache_file_lock = FileLock(cache_filename + ".lock", timeout=self.CACHE_FILE_LOCK_TIMEOUT)
-        try:
-            cache_file_lock.acquire()
-        finally:
-            # We make an assumption here that if we can obtain the lock, no one will
-            # be trying to write to the file anymore, so it should be safe to release the lock
-            # before reading.
-            cache_file_lock.release()
         with open(cache_filename, "r") as cache_file:
-            for instance in self._multi_worker_islice(cache_file, self.deserialize_instance):
-                yield instance
+            yield from self._multi_worker_islice(cache_file, self.deserialize_instance)
 
     def _instances_to_cache_file(self, cache_filename, instances) -> None:
-        with FileLock(cache_filename + ".lock", timeout=self.CACHE_FILE_LOCK_TIMEOUT):
+        # We serialize to a temp file first in case anything goes wrong while
+        # writing to cache (e.g., the computer shuts down unexpectedly).
+        # Then we just copy the file over to `cache_filename`.
+        with tempfile.NamedTemporaryFile("w+") as temp_file:
+            logger.info("Caching instances to temp file %s", temp_file.name)
+            for instance in Tqdm.tqdm(instances):
+                temp_file.write(self.serialize_instance(instance) + "\n")
+
+            # we are copying the file before closing it, so flush to avoid truncation
+            temp_file.flush()
+            # shutil.copyfileobj() starts at the current position, so go to the start
+            temp_file.seek(0)
+
+            logger.info("copying %s to cache at %s", temp_file.name, cache_filename)
             with open(cache_filename, "w") as cache:
-                for instance in Tqdm.tqdm(instances):
-                    cache.write(self.serialize_instance(instance) + "\n")
+                shutil.copyfileobj(temp_file, cache)
```

Comment:
To do this right, you have to _move_/_rename_ the file. `copyfileobj` will copy the file one buffer at a time, and you could still end up with a truncated file if the power goes out.

This code does this: https://github.com/allenai/datastore/blob/master/python/datastore.py#L200
Though it's not very clear in that snippet, it makes sure that the temporary file and the final destination file are in the same directory (or at least on the same filesystem). Otherwise, that call to `Path().rename()` is not atomic (or it will fail, I don't remember which).

Reply:
That's a good point. We should update our `file_utils` caching system as well then.